### PR TITLE
Add `compute_loss` method to `Model`

### PR DESCRIPTION
**Pull request by arxyzan**
This PR adds `compute_loss` method as the primary method for calculating loss. 

We made this decision since the models with the same task do not necessarily follow the same pattern for computing loss. That's why most libraries compute loss in their forward method. Although we took a slighly different approach. Instead of messing up all forward methods with an if statement for computing loss which is inaccessible for the end user, we added a seperate method for this purpose.
This method takes in two main inputs as `inputs` and `targets`. Subclasses can override this notation.

### CODE REVIEW

File: hezar/models/image2text/beit_roberta/beit_roberta_image2text.py [hunk 1]
```diff
@@ -65,14 +64,18 @@ def forward(
             encoder_outputs=encoder_outputs,
             past_key_values=past_key_values,
             decoder_inputs_embeds=decoder_inputs_embeds,
-            labels=labels,
             use_cache=use_cache,
```
Comment (pooya-mohammadi) [inline hezar/models/image2text/beit_roberta/beit_roberta_image2text.py:67]: It's better to explicitly set labels to None! So code reviewers will get our approach.

File: hezar/models/image2text/crnn/crnn_image2text.py [hunk 2]
```diff
@@ -53,6 +54,20 @@ def forward(self, pixel_values, **kwargs):
         x = nn.functional.log_softmax(x, 2)
         return x
 
+    def compute_loss(self, logits: torch.Tensor, labels: torch.Tensor, labels_lengths: torch.Tensor = None):
+        if labels_lengths is None:
```
Comment (pooya-mohammadi) [inline hezar/models/image2text/crnn/crnn_image2text.py:58]: If labels_lengths should be None, why the default value is None?!

File: hezar/models/image2text/crnn/crnn_image2text.py [hunk 3]
```diff
@@ -53,6 +54,20 @@ def forward(self, pixel_values, **kwargs):
         x = nn.functional.log_softmax(x, 2)
         return x
 
+    def compute_loss(self, logits: torch.Tensor, labels: torch.Tensor, labels_lengths: torch.Tensor = None):
```
Comment (pooya-mohammadi) [inline hezar/models/image2text/crnn/crnn_image2text.py:57]: Adding func's output annotation is a plus :D

File: hezar/models/image2text/beit_roberta/beit_roberta_image2text.py (hunk 1 above)
Comment (arxyzan) [reply hezar/models/image2text/beit_roberta/beit_roberta_image2text.py:68]: @pooya-mohammadi Nice catch. Done.

**Review by pooya-mohammadi — CHANGES_REQUESTED** [review]
image
@arxyzan jan creating an object each time in compute loss and garbage collecting it reduces the speed. 
On top of that users may want to add `weight` or other types loss function inputs to computer_loss which makes it more challenging to work with. 
I think it's better to create the loss function in the init for once and use it afterward.

**Comment by arxyzan** [comment]
> image @arxyzan jan creating an object each time in compute loss and garbage collecting it reduces the speed. On top of that users may want to add `weight` or other types loss function inputs to computer_loss which makes it more challenging to work with. I think it's better to create the loss function in the init for once and use it afterward.

@pooya-mohammadi You're right. I thought that having criterion as `self.criterion` which is a `nn.Module`, would register it in the state dict which it doesn't! (I just checked). 

But there's another issue here. If the models require the users to only implement `compute_loss` for their own custom models, they'd do all they actually need inside one method. How do think we should implement such thing in a way that it doesn't complicate working with custom models in Hezar? If there's no way around, we should put criterions in all the `__init__` methods.

File: hezar/models/image2text/crnn/crnn_image2text.py (hunk 2 above)
Comment (arxyzan) [reply hezar/models/image2text/crnn/crnn_image2text.py:58]: @pooya-mohammadi Hmm, I don't remember exactly but I just tried to remove it and got warning (does not match signature ...) from the IDE.

**Comment by pooya-mohammadi** [comment]
> > image @arxyzan jan creating an object each time in compute loss and garbage collecting it reduces the speed. On top of that users may want to add `weight` or other types loss function inputs to computer_loss which makes it more challenging to work with. I think it's better to create the loss function in the init for once and use it afterward.
> 
> @pooya-mohammadi You're right. I thought that having criterion as `self.criterion` which is a `nn.Module`, would register it in the state dict which it doesn't! (I just checked).
> 
> But there's another issue here. If the models require the users to only implement `compute_loss` for their own custom models, they'd do all they actually need inside one method. How do think we should implement such thing in a way that it doesn't complicate working with custom models in Hezar? If there's no way around, we should put criterions in all the `__init__` methods.

@arxyzan jan, users still can do all they want in a single method if they want to. They can create the loss function inside the method and use it! If they were pro-programmers they could extend the init method and overwrite the criterion. 
All options are available for users.

**Comment by arxyzan** [comment]
> > > image @arxyzan jan creating an object each time in compute loss and garbage collecting it reduces the speed. On top of that users may want to add `weight` or other types loss function inputs to computer_loss which makes it more challenging to work with. I think it's better to create the loss function in the init for once and use it afterward.
> > 
> > 
> > @pooya-mohammadi You're right. I thought that having criterion as `self.criterion` which is a `nn.Module`, would register it in the state dict which it doesn't! (I just checked).
> > But there's another issue here. If the models require the users to only implement `compute_loss` for their own custom models, they'd do all they actually need inside one method. How do think we should implement such thing in a way that it doesn't complicate working with custom models in Hezar? If there's no way around, we should put criterions in all the `__init__` methods.
> 
> @arxyzan jan, users still can do all they want in a single method if they want to. They can create the loss function inside the method and use it! If they were pro-programmers they could extend the init method and overwrite the criterion. All options are available for users.

@pooya-mohammadi Done. I did it in a neat way! ;)

File: hezar/models/image2text/crnn/crnn_image2text.py (hunk 2 above)
Comment (arxyzan) [reply hezar/models/image2text/crnn/crnn_image2text.py]: @pooya-mohammadi Done.

**Review by pooya-mohammadi — APPROVED** [review]
